reference-implementation/lib/readable-byte-stream.js

```diff
@@ -0,0 +1,965 @@
+    }
+
+    return new ReadableByteStreamReader(this);
+  }
+}
+
+class ReadableByteStreamController {
+  constructor(controlledReadableByteStream, underlyingByteSource) {
+    if (IsReadableByteStream(controlledReadableByteStream) === false) {
+      throw new TypeError('ReadableByteStreamController can only be constructed with a ReadableByteStream instance');
+    }
+
+    if (controlledReadableByteStream._controller !== undefined) {
+      throw new TypeError(
+          'ReadableByteStreamController instances can only be created by the ReadableByteStream constructor');
+    }
+
+    this._controlledReadableByteStream = controlledReadableByteStream;
+
+    this._underlyingByteSource = underlyingByteSource;
+
+    this._considerReissueUnderlyingByteSourcePullOrPullInto = false;
+    this._insideUnderlyingByteSource = false;
+
+    this._pendingPullIntos = [];
+
+    this._queue = [];
+    this._totalQueuedBytes = 0;
+
+    InvokeOrNoop(underlyingByteSource, 'start', [this]);
+  }
+
+  close() {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.close can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._closeRequested) {
+      throw new TypeError('The stream has already been closed; do not close it again!');
+    }
+    if (stream._state !== 'readable') {
+      throw new TypeError('The stream is not in the readable state and cannot be closed');
+    }
+
+    if (this._totalQueuedBytes > 0) {
+      this._closeRequested = true;
+
+      return;
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined || IsReadableByteStreamReader(reader)) {
+      CloseReadableByteStream(stream);
+      return;
+    }
+
+    assert(IsReadableByteStreamByobReader(reader), 'reader must be ReadableByteStreamByobReader');
+
+    if (this._pendingPullIntos.length > 0 && this._pendingPullIntos[0].bytesFilled > 0) {
+      DestroyReadableByteStreamController(this);
+      const e = new TypeError('Insufficient bytes to fill elements in the given buffer');
+      ErrorReadableByteStream(stream, e);
+
+      throw e;
+    }
+
+    CloseReadableByteStream(stream);
+  }
+
+  enqueue(chunk) {
+    if (!IsReadableByteStreamController(this)) {
+      throw new TypeError(
+          'ReadableByteStreamController.prototype.enqueue can only be used on a ReadableByteStreamController');
+    }
+
+    const stream = this._controlledReadableByteStream;
+
+    if (stream._closeRequested) {
+      throw new TypeError('stream is closed or draining');
+    }
+    if (stream._state !== 'readable') {
+      throw new TypeError('The stream is not in the readable state and cannot be enqueued to');
+    }
+
+    const reader = stream._reader;
+
+    if (reader === undefined) {
+      EnqueueInReadableByteStreamController(this, chunk);
+      return;
```

Comment:
IMO most of the early returns in this function would actually be simpler as if/elses. "If there's no reader, do X. Otherwise, do Y." Instead of "If there's no reader, do X and return. Do Y."

Just a style thing though, not important.